### DISCUSSION TITLE
HTTP/1 protocol error: peer closed connection without sending complete message body

### DISCUSSION THREAD

**Discussion by erikshestopal** [mitmproxy/mitmproxy#6253] [Q&A]
The setup (`mitm v9`): 
1. Client (in this case an application running inside a K8s pod) sends an HTTP request to `https://example.com` via our a custom MITM proxy server running in `regular` mode. 
2. For every outbound request, we will insert a separate IP rotating proxy into the flow, such that the entire cycle looks something like this: Client -> MITM Proxy -> IP Rotating Proxy -> Upstream Server. 
3. Depending on the hostname and certain rate-limiting rules, the MITM proxy server will either allow the request to pass through or deny with a 429 status code with a `Retry-After` header. 
4. If the client application receives a 429 status response, it will wait for a duration specified by the `Retry-After` header before attempting the same request again. 

The problem I'm noticing is that upon receiving a 429 status code from the proxy and attempting to subsequently retry the request, the connection hangs until the watchdog timer hits 10 minutes before erroring out with the message below. This error does not seem to occur if I don't set the 429 status code on the flow. So far, I've only noticed this with `POST` requests that have binary data like documents in the request body. 

```
HTTP/1 protocol error: peer closed connection without sending complete message body (received 0 bytes, expected 131746)
```

The debug mode logs are roughly equivalent to the screenshot below but happy to provide any additional logging output. 

<img width="1684" alt="Screenshot 2023-07-13 at 3 48 09 PM" src="https://github.com/mitmproxy/mitmproxy/assets/6404663/e67e2168-c42a-4cfb-996b-185fb4088b0d">

Would appreciate any insights on how to get to the root cause. Thanks!

**Comment by mhils** [comment]
> Would appreciate any insights on how to get to the root cause.

Simplify your setup and make it reproducible. :)

The log needs more context, there are likely (Disconnect) events queued up that cause the error.

**Reply by erikshestopal** [reply]
@mhils Trying to reproduce locally is what I've been attempting to due the last week haha. 

What additional log context would be helpful? And when you say there are Disconnect events queued up, what am I looking for specifically in the logs?

**Reply by mhils** [reply]
The log excerpt you posted is basically just the last event bubbling back up, so pretty much all context is missing here. Disconnect events would be that - something like ClientDisconnected appearing in the connection.

**Reply by erikshestopal** [reply]
Not seeing any of the ClientDisconnected events but what I did find interesting is the logs in the screenshot, where the event chain just seems to disappear for like 50(?!) seconds before we get a disconnect from out intermediate proxy. Any ideas here? I know I'm asking for the impossible 😓 

<img width="750" alt="Screenshot 2023-07-14 at 12 24 51 PM" src="https://github.com/mitmproxy/mitmproxy/assets/6404663/5e0c7b95-5398-42bd-8dd4-b4a2a1f822c3">

**Reply by mhils** [reply]
This screenshot is more useful already. Here we see the server eventually closing a connection after ~50s of inactivity. That's not unexpected.
Whatsoever, the expectation here is 1) you provide something reproducible and/or full logs, or 2) reach out privately for (paid) consulting to someone on the team. :)

**Reply by erikshestopal** [reply]
@mhils How can I reach your team privately?

**Comment by mhils** [comment]
Feel free to reach out via the email address on my website. :)

**Reply by erikshestopal** [reply]
Sent.
